### PR TITLE
chore(chart): helm chart for dynamic-nfs-provisioner 

### LINKED ISSUE #15: creating chart for the dynamic-nfs-provisioner
**Describe the bug:** Creating chart for the dynamic-nfs-provisioner  to make the deployment more generic

**Expected behaviour:**  Deployment of dynamic-nfs-provisioner with helm

**Environment details:**
- others:

### PR DESCRIPTION
**Pull request by rahulkrishnanfs**
Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

## Pull Request template

**Why is this PR required? What issue does it fix?**:
Chart for the dynamic-nfs-provisioner to make the deployment more generic
**What this PR does?**:
 Helm chart for dynamic-nfs-provisioner
**Does this PR require any upgrade changes?**:
No
**If the changes in this PR are manually verified, list down the scenarios covered:**:
Install the chart to the desired namespace using helm 

**Checklist:**
- [x] Fixes #15 
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

**Comment by niladrih** [comment]
@rahulkrishnanfs -- I think we may need the dynamic-localpv-provisioner chart as a dependent chart here because we are using `openebs-hostpath` as the default backend storageClass.

**Review by niladrih — CHANGES_REQUESTED** [review]
@rahulkrishnanfs -- I think we may need the dynamic-localpv-provisioner chart as a dependent chart here because we are using openebs-hostpath as the default backend storageClass.

**Comment by niladrih** [comment]
@rahulkrishnanfs -- Tested this, I got this error from the events of the application pod. Pod remained in ContainerCreating state.
```yaml
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8 --scope -- mount -t nfs 10.100.81.23:/ /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8
Output: Running scope as unit: run-r99ee95a41b244866ac61875bf8591202.scope
mount: /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8: bad option; for several filesystems (e.g. nfs, cifs) you might need a /sbin/mount.<type> helper program.
  Warning  FailedMount  30s  kubelet  MountVolume.SetUp failed for volume "pvc-a384f7b7-236a-4508-86a7-6c59359664d8" : mount failed: exit status 32
Mounting command: systemd-run
```

This is the storageClass that got created by default (`kubectl get sc openebs-kernel-nfs -o yaml`)
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    cas.openebs.io/config: |
      - name: NFSServerType
        value: kernel
    meta.helm.sh/release-name: nfspv-provisioner
    meta.helm.sh/release-namespace: openebs
    openebs.io/cas-type: nfsrwx
  creationTimestamp: "2021-04-16T16:00:07Z"
  labels:
    app.kubernetes.io/managed-by: Helm
  name: openebs-kernel-nfs
  resourceVersion: "277852"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/openebs-kernel-nfs
  uid: 04befa04-44e8-45b8-aac2-54cffc6bc21b
provisioner: openebs.io/nfsrwx
reclaimPolicy: Delete
volumeBindingMode: Immediate
```
I used the this command to apply the chart when `pwd` was dynamic-nfs-provisioner/deploy/charts/nfspv-provisioner
`helm install -n openebs nfspv-provisioner .`

Might be because of the absence of BackendStorageClass?
cc: @kmova 

EDIT: This is due to https://github.com/openebs/dynamic-nfs-provisioner/issues/16. Testing environment was not configured correctly.

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Thank you for installing {{ .Chart.Name }} 😀
+
+Your release is named {{ .Release.Name }} and it's installed to namespace: {{ .Release.Namespace }}.
+
+The OpenEBS NFSPV Provisioner has been installed check its status by running:
+$ kubectl get pods -n {{ .Release.Namespace }}
+
+Use `kubectl get bd -n {{ .Release.Namespace }} ` to see the list of
```
Comment (niladrih) [inline deploy/charts/nfspv-provisioner/templates/NOTES.txt]: This command does not seem relevant to this chart.

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+blockdevices attached to the Kubernetes cluster nodes.
+
+For more information, visit our Slack at https://openebs.io/community or view
+the documentation online at http://docs.openebs.io/.
```
Comment (niladrih) [inline deploy/charts/nfspv-provisioner/templates/NOTES.txt]: NFS provisioner docs are not available at docs.openebs.io at the moment. Could we use https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/README.md instead?

**Review by niladrih — CHANGES_REQUESTED** [review]
@rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment. 

The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.

Suggested some chart deploy message changes.

**Comment by quickerdevops** [comment]
> @rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment.
> 
> The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.
> 
> Suggested some chart deploy message changes.

@niladrih  Thanks mate for the suggestion. I have changed the directory structure as you suggested

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt (hunk 1 above)
Comment (quickerdevops) [reply deploy/charts/nfspv-provisioner/templates/NOTES.txt]: Thanks mate , removed it

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt (hunk 2 above)
Comment (quickerdevops) [reply deploy/charts/nfspv-provisioner/templates/NOTES.txt]: Pointed to the developer readme file as you suggested , Thanks

**Comment by rahulkrishnanfs** [comment]
> @rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment.
> 
> The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.
> 
> Suggested some chart deploy message changes.

@niladrih Thanks mate for the suggestion. I have changed the directory structure as you suggested

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt (hunk 2 above)
Comment (rahulkrishnanfs) [reply deploy/charts/nfspv-provisioner/templates/NOTES.txt]: Thanks mate , Updated to the developer readme file as you suggested

File: deploy/charts/nfspv-provisioner/templates/NOTES.txt (hunk 1 above)
Comment (rahulkrishnanfs) [reply deploy/charts/nfspv-provisioner/templates/NOTES.txt]: Removed the mentioned line, thanks

File: deploy/helm/charts/nfspv-provisioner/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+
+## Configuration
+
+The following table lists the configurable parameters of the OpenEBS NFSPV Provisioner chart and their default values.
+
+| Parameter                                   | Description                                   | Default                                   |
+| ------------------------------------------- | --------------------------------------------- | ----------------------------------------- | 
+| `analytics.enabled`                         | Enable sending stats to Google Analytics          | `true`                          |
+| `imagePullSecrets`                          | Provides image pull secrect                       | `""`                            |
```
Comment (niladrih) [inline deploy/helm/charts/nfspv-provisioner/README.md]:
```suggestion
| `imagePullSecrets`                          | Provides image pull secret                       | `""`                            |
```

File: deploy/helm/charts/nfspv-provisioner/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+| `nfspv.enabled`                             | Enable NFSPV Provisioner                          | `true`                          |
+| `nfspv.image.registry`                      | Registry for NFSPV Provisioner image              | `""`                            |
+| `nfspv.image.repository`                    | Image repository for NFSPV Provisioner            | `openebs/provisioner-nfs-amd64` |
+| `nfspvpv.image.tag`                         |	Image tag for NFSPV Provisioner	                  | `0.2.0`                         |
```
Comment (niladrih) [inline deploy/helm/charts/nfspv-provisioner/README.md]:
```suggestion
| `nfspv.image.tag`                           | Image tag for NFSPV Provisioner	                  | `0.2.0`                         |
```

File: deploy/helm/charts/nfspv-provisioner/templates/kernel-nfs-storageclass.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#Storage classes for OpenEBS NFS Dynamic PV
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: {{ .Values.nfsStorageClass.name }}
+  annotations:
+    openebs.io/cas-type: nfsrwx
+    cas.openebs.io/config: |
+      - name: NFSServerType
+        value: {{ .Values.nfsStorageClass.nfsServerType }}
```
Comment (niladrih) [inline deploy/helm/charts/nfspv-provisioner/templates/kernel-nfs-storageclass.yaml:10]:
```suggestion
        value: {{ .Values.nfsStorageClass.nfsServerType }}
{{- if .Values.nfsStorageClass.BackendStorageClass }}
      - name: BackendStorageClass
        value: {{ .Values.nfsStorageClass.BackendStorageClass }}
{{- end }}
```
Can we do something like this, so that the user can specify the backend storageClass with a flag like `--set nfsStorageClass.BackendStorageClass=<storageClass>` ?

File: deploy/helm/charts/nfspv-provisioner/values.yaml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  tolerations: [ ]
+  affinity: { }
+  healthCheck:
+    initialDelaySeconds: 30
+    periodSeconds: 60
+
+nfsStorageClass:
+  name: openebs-kernel-nfs
+  reclaimPolicy: Delete
+  nfsServerType: kernel
```
Comment (niladrih) [inline deploy/helm/charts/nfspv-provisioner/values.yaml:66]:
```suggestion
  nfsServerType: kernel
  BackendStorageClass: ""
```
Could we add the option here to add a BackendStorageClass?

**Review by niladrih — CHANGES_REQUESTED** [review]

File: deploy/helm/charts/nfspv-provisioner/values.yaml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  name: nfspv-provisioner
+  enabled: true
+  annotations: {}
+  podLabels:
+    name: openebs-nfspv-provisioner
+  image:
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry:
+    repository: openebs/provisioner-nfs-amd64
```
Comment (mynktl) [inline deploy/helm/charts/nfspv-provisioner/values.yaml]: registry should be `openebs/` and repository `provisioner-nfs-amd64`

File: deploy/helm/charts/nfspv-provisioner/README.md (hunk 4 above)
Comment (mynktl) [reply deploy/helm/charts/nfspv-provisioner/README.md]: typo: `nfspvpv` -> `nfspv`

File: deploy/helm/charts/nfspv-provisioner/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+apiVersion: v2
+name: nfspv-provisioner
+description: Helm chart for OpenEBS Dynamic NFS PV. For instructions to install OpenEBS Dynamic NFS PV using helm chart, refer to https://github.com/openebs/dynamic-nfs-provisioner.
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/Chart.yaml]:
```suggestion
description: Helm chart for OpenEBS Dynamic NFS PV. For instructions to install OpenEBS Dynamic NFS PV using helm chart, refer to https://openebs.github.io/dynamic-nfs-provisioner.
```

File: deploy/helm/charts/nfspv-provisioner/templates/NOTES.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+Thank you for installing {{ .Chart.Name }} 😀
+
+Your release is named {{ .Release.Name }} and it's installed to namespace: {{ .Release.Namespace }}.
+
+The OpenEBS NFSPV Provisioner has been installed check its status by running:
+$ kubectl get pods -n {{ .Release.Namespace }}
+
+For more information, visit our Slack at https://openebs.io/community or view
+the documentation online at https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/develop/developer.md .
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/templates/NOTES.txt]:
```suggestion
the documentation online at https://github.com/openebs/dynamic-nfs-provisioner/.
```

File: deploy/helm/charts/nfspv-provisioner/values.yaml [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+nfspv:
+  name: nfspv-provisioner
+  enabled: true
+  annotations: {}
+  podLabels:
+    name: openebs-nfspv-provisioner
+  image:
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry:
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/values.yaml]:
```suggestion
    registry: openebs
```

File: deploy/helm/charts/nfspv-provisioner/values.yaml (hunk 7 above)
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/values.yaml]:
```suggestion
    repository: openebs/provisioner-nfs
```

File: deploy/helm/charts/nfspv-provisioner/values.yaml (hunk 6 above)
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/values.yaml:66]:
```suggestion
  nfsServerType: kernel
  backendStorageClass: ""
```

File: deploy/helm/charts/nfspv-provisioner/templates/kernel-nfs-storageclass.yaml (hunk 5 above)
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/templates/kernel-nfs-storageclass.yaml:10]:
```suggestion
        value: {{ .Values.nfsStorageClass.nfsServerType }}
{{- if .Values.nfsStorageClass.backendStorageClass }}
      - name: BackendStorageClass
        value: {{ .Values.nfsStorageClass.backendStorageClass }}
{{- end }}
```

File: deploy/helm/charts/nfspv-provisioner/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
+apiVersion: v2
+name: nfspv-provisioner
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/Chart.yaml:2]:
```suggestion
name: nfs-provisioner
```

File: deploy/helm/charts/nfspv-provisioner/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+| akhilerm | akhil.mohan@mayadata.io |  |
+| kiranmova | kiran.mova@mayadata.io |  |
+| prateekpandey14 | prateek.pandey@mayadata.io |  |
+| rahulkrishnanra | rahulkrishnanfs@gmail.com |  |
+
+
+## Get Repo Info
+
+```console
+helm repo add openebs-nfspv https://openebs.github.io/dynamic-nfspv-provisioner
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/README.md:23]:
```suggestion
helm repo add openebs-nfs https://openebs.github.io/dynamic-nfs-provisioner
```

File: deploy/helm/charts/nfspv-provisioner/README.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+_See [helm repo](https://helm.sh/docs/helm/helm_repo/) for command documentation._
+
+## Install Chart
+
+Please visit the [link](https://openebs.github.io/dynamic-nfspv-provisioner/) for install instructions via helm3.
+
+```console
+# Helm
+$ helm install [RELEASE_NAME] openebs-nfspv/nfspv-provisioner
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/README.md:35]:
```suggestion
$ helm install [RELEASE_NAME] openebs-nfs/nfs-provisioner
```

File: deploy/helm/charts/nfspv-provisioner/README.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+| `nfspv.enableLeaderElection`                | Enable leader election                            | `true`                          |
+| `rbac.create`                               | Enable RBAC Resources                             | `true`                          |
+| `rbac.pspEnabled`                           | Create pod security policy resources              | `false`                         |
+| `nfspv.affinity`                            | NFSPV Provisioner pod affinity                    | `{}`                            | 
+Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`.
+
+Alternatively, a YAML file that specifies the values for the parameters can be provided while installing the chart. For example,
+
+```bash
+helm install <release-name> -f values.yaml ----namespace openebs nfspv-provisioner
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/README.md:90]:
```suggestion
helm install <release-name> -f values.yaml ----namespace openebs nfs-provisioner
```

**Review by niladrih — APPROVED** [review]
Looks good to me. Some changes will be taken up in a separate PR.

File: deploy/helm/charts/nfspv-provisioner/values.yaml [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+nfspv:
+  name: nfspv-provisioner
+  enabled: true
+  annotations: {}
+  podLabels:
+    name: openebs-nfspv-provisioner
+  image:
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry: openebs
```
Comment (kmova) [inline deploy/helm/charts/nfspv-provisioner/values.yaml]:
```suggestion
    registry: 
```

**Review by niladrih — APPROVED** [review]
LGTM
